test/session-pool.js

```diff
@@ -16,2022 +16,1575 @@
+          create: () => Promise.reject('Error'),
+          delete: () => {
+            deleteCalled = true;
+            return Promise.resolve();
+          },
+        };
+
+        sessionPool.session_ = function() {
+          return fakeSession;
+        };
 
-    it('should emit a close event', function(done) {
-      sessionPool.on('close', done);
-      sessionPool.close();
+        return sessionPool.createWriteSession().catch(error => {
+          assert.strictEqual(error, 'Error');
+          assert.strictEqual(deleteCalled, true);
+        });
+      });
     });
 
-    it('should emit an empty event', function(done) {
-      sessionPool.on('empty', done);
-      sessionPool.close();
+    describe('open', function() {
+      it('should set isOpen to true', function() {
+        return sessionPool.close().then(function() {
+          sessionPool = new SessionPool(DATABASE, {minReads: 4, minWrites: 4});
+          sessionPool.session_ = function() {
+            return new Object({
+              create: () => Promise.resolve(),
+              delete: () => Promise.resolve(),
+            });
+          };
+          sessionPool.open();
+          assert.strictEqual(sessionPool.isOpen, true);
+          assert.strictEqual(sessionPool.getStats().readPool.size, 4);
+          assert.strictEqual(sessionPool.getStats().writePool.size, 0);
+          clearTimeout(sessionPool.pingTimeoutHandle);
+          return sessionPool.close();
+        });
+      });
     });
 
-    it('should call stopHouseKeeping_', function(done) {
-      sessionPool.stopHouseKeeping_ = done;
-      sessionPool.close();
+    describe('close', function() {
+      it('should destroy all sessions', function() {
+        return sessionPool.close().then(function() {
+          sessionPool = new SessionPool(DATABASE, {minReads: 4, minWrites: 4});
+          sessionPool.createTransaction_ = () => Promise.resolve();
+          sessionPool.session_ = function() {
+            return new Object({
+              create: () => Promise.resolve(),
+              delete: () => Promise.resolve(),
+            });
+          };
+          sessionPool.open();
+          assert.strictEqual(sessionPool.isOpen, true);
+          assert.strictEqual(sessionPool.getStats().readPool.size, 4);
+          assert.strictEqual(sessionPool.getStats().writePool.size, 0);
+          clearTimeout(sessionPool.pingTimeoutHandle);
+          return sessionPool.close().then(() => {
+            assert.strictEqual(sessionPool.isOpen, false);
+            assert.strictEqual(sessionPool.getStats().readPool.size, 0);
+            assert.strictEqual(sessionPool.getStats().writePool.size, 0);
+          });
+        });
+      });
     });
 
-    it('should destroy all sessions', function() {
-      var fakeReads = [{}, {}];
-      var fakeWrites = [{}];
-      var fakeBorrowed = [{}, {}, {}];
+    describe('getReadSession and getWriteSession with minReads and minWrites', function() {
```

Comment:
Can we break this into multiple describe blocks?

Reply:
updated